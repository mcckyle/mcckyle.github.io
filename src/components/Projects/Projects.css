/* Filename: Projects.css */

/* ====================================
   Projects Styles.
======================================= */
#projects .section-title
{
    color: var(--accent);
}

.projects-list
{
    display: grid;
    grid-template-columns: repeat(auto-fit, minmax(300px, 1fr));
    gap: 2.5rem;

    margin-top: 2rem;
    padding: 0;
    list-style: none;
}

.project-title
{
    font-size: 1.35rem;
    font-weight: 600;
    color: var(--primary);
    margin-bottom: 0.75rem;
}

/* Project Description Styles. */
.project-description
{
    color: var(--text);
    line-height: 1.65;
    margin-bottom: 1.25rem;
}

/* Project Tech Tags Styles. */
.project-tech
{
    display: flex;
    flex-wrap: wrap;
    gap: 0.5rem;
    padding: 0;
    margin: 0;
    list-style: none;
}

.tech-tag
{
    background: color-mix(in oklab, var(--accent), white 88%);
    color: var(--accent);
    font-size: 0.85rem;
    font-weight: 500;
    padding: 0.35rem 0.65rem;
    border-radius: 0.5rem;
    letter-spacing: 0.3px;

    transition: background var(--transition), color var(--transition);
}

.tech-tag:hover
{
    background: var(--accent);
    color: #fff;
}
