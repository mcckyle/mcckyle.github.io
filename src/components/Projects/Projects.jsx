//Filename: Projects.jsx
//Author: Kyle McColgan
//Date: 28 October 2025
//Description: This file contains the component for the personal React project projects section.

import React from "react";
import "./Projects.css";

function Projects()
{
    const projects = [
      {
        title: "Automated Snapshot Management Script",
        description:
          "A Bash script that automates system snapshot storage with Snapper. Enhances reliability by managing snapshots efficiently and transferring them to local storage to conserve disk space while maintaining version history.",
        tech: ["Bash", "Linux", "Snapper"],
      },
      {
        title: "ShowMeTasks",
        description:
          "A full-stack to-do list application built with Java, React, and Spring Boot. Offers a seamless task management experience with responsive design, robust backend logic, and secure RESTful APIs.",
        tech: ["Java", "React", "Spring Boot", "MySQL"],
      },
    ];

    return (
        <div id="projects" role="region" aria-labelledby="projects-title" className="section-content">
          <h2 id="projects-title" className="section-title">Projects</h2>
          <ul className="projects-list">
            {projects.map((project, index) => (
              <li key={index} className="card">
                <h3 className="project-title">{project.title}</h3>
                <p className="project-description">{project.description}</p>
                <ul className="project-tech">
                  {project.tech.map((tech, i) => (
                    <li key={i} className="tech-tag">
                      {tech}
                    </li>
                  ))}
                </ul>
              </li>
            ))}
          </ul>
       </div>
    );
}

export default Projects;
