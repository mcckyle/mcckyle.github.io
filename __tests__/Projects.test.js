//Filename: Projects.test.js
//Author: Kyle McColgan
//Date: 2 October 2025
//Description: This file contains the Jest unit tests for the personal React project projects section.

import React from 'react';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import Projects from '../src/components/Projects/Projects.jsx';

describe('Projects component', () => {
    beforeEach(() => {
        render(<Projects />);
    });

    //Test #1
    it('renders the Projects section', () => {
        const section = screen.getByRole('region', { name: /Projects/i });
        expect(section).toBeInTheDocument();
    });

    //Test #2
    it('renders the section title correctly', () => {
        const title = screen.getByRole('heading', { level: 2, name: /Projects/i });
        expect(title).toHaveTextContent('Projects');
    });

    //Test #3
    it('renders the list of project cards', () => {
        const projectCards = screen.getAllByRole('listitem', { name: '' }); //Default list items.
        expect(projectCards).toHaveLength(2 + 3 + 4); //2 cards, 3 tech tags (Proj. 1), 4 tech tags (Proj. 2).
    });

    //Test #4
    it('renders the first project correctly', () => {
        const firstTitle = screen.getByText(/Automated Snapshot Management Script/i);
        const firstDesc = screen.getByText(/A Bash Script designed to automate/i);
        expect(firstTitle).toBeInTheDocument();
        expect(firstDesc).toBeInTheDocument();

        // const firstTechTags = screen.getAllByText(/Bash|Linux|Snapper/i);
        // expect(firstTechTags).toHaveLength(3);
    });

    //Test #5
    it('renders the second project correctly', () => {
        const secondTitle = screen.getByText(/ShowMeTasks/i);
        const secondDesc = screen.getByText(/A full-stack to-do list application/i);
        expect(secondTitle).toBeInTheDocument();
        expect(secondDesc).toBeInTheDocument();

        // const secondTechTags = screen.getAllByText(/Java|React|Spring Boot|MySQL/i);
        // expect(secondTechTags).toHaveLength(4);
    });

    //Test #6
    // it('renders each project title in bold', () => {
    //     const boldElements = screen.getAllByRole('strong');
    //     expect(boldElements.length).toBeGreaterThanOrEqual(2);
    // });

    //Test #7
    it('does not render any empty project cards', () => {
        const projectCards = screen.getAllByRole('listitem', { name: '' }); //Default list items.
        projectCards.forEach(card => {
            expect(card).not.toBeEmptyDOMElement();
        });
    });

    //Test #8
    // it('uses a <ul> to list projects', () => {
    //     const list = screen.getByRole('list');
    //     expect(list.tagName).toBe('UL');
    // });

    //Test #9
    it('renders without console errors', () => {
        const consoleError = jest.spyOn(console, 'error').mockImplementation();
        expect(consoleError).not.toHaveBeenCalled();
        consoleError.mockRestore();
    });

    //Test #10
    it('matches the snapshot of the Projects section', () => {
        const { asFragment } = render(<Projects />);
        expect(asFragment()).toMatchSnapshot();
    });
});
