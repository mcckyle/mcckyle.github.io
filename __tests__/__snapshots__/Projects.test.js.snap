// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Projects component matches the snapshot of the Projects section 1`] = `
<DocumentFragment>
  <div
    aria-labelledby="projects-title"
    class="section-content"
    id="projects"
    role="region"
  >
    <h2
      class="section-title"
      id="projects-title"
    >
      Projects
    </h2>
    <ul
      class="projects-list"
    >
      <li
        class="card"
      >
        <h3
          class="project-title"
        >
          Automated Snapshot Management Script
        </h3>
        <p
          class="project-description"
        >
          A Bash script that automates system snapshot storage with Snapper. Enhances reliability by managing snapshots efficiently and transferring them to local storage to conserve disk space while maintaining version history.
        </p>
        <ul
          class="project-tech"
        >
          <li
            class="tech-tag"
          >
            Bash
          </li>
          <li
            class="tech-tag"
          >
            Linux
          </li>
          <li
            class="tech-tag"
          >
            Snapper
          </li>
        </ul>
      </li>
      <li
        class="card"
      >
        <h3
          class="project-title"
        >
          ShowMeTasks
        </h3>
        <p
          class="project-description"
        >
          A full-stack to-do list application built with Java, React, and Spring Boot. Offers a seamless task management experience with responsive design, robust backend logic, and secure RESTful APIs.
        </p>
        <ul
          class="project-tech"
        >
          <li
            class="tech-tag"
          >
            Java
          </li>
          <li
            class="tech-tag"
          >
            React
          </li>
          <li
            class="tech-tag"
          >
            Spring Boot
          </li>
          <li
            class="tech-tag"
          >
            MySQL
          </li>
        </ul>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;
